File: src/main/java/com/github/mgramin/sqlboot/resource_type/ResourceType.java
/**
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2016-2017 Maksim Gramin
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.resource_type;

import java.util.List;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.mgramin.sqlboot.actions.generator.ActionGenerator;
import com.github.mgramin.sqlboot.exceptions.BootException;
import com.github.mgramin.sqlboot.resource.DbResource;
import com.github.mgramin.sqlboot.model.IDbResourceCommand;
import com.github.mgramin.sqlboot.uri.Uri;

/**
 * Created by maksim on 16.05.17.
 */
public interface ResourceType {

    /**
     *
     * @return
     */
    @JsonProperty
    String name();

    /**
     *
     * @return
     */
    List<String> aliases();

    /**
     *
     * @param uri
     * @return
     * @throws BootException
     */
    List<DbResource> read(Uri uri) throws BootException;

    @Deprecated
    default List<ActionGenerator> generators() {
        return null;
    }

    @Deprecated
    default List<DbResource> read(Uri uri, @Deprecated IDbResourceCommand command, @Deprecated String aggregatorName) throws BootException {
        return null;
    }

}

File: src/main/kotlin/com/github/mgramin/sqlboot/model/resourcetype/ResourceType.kt
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2019, CROC Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.github.mgramin.sqlboot.model.resourcetype

import com.fasterxml.jackson.databind.JsonNode
import com.github.mgramin.sqlboot.model.resource.DbResource
import com.github.mgramin.sqlboot.model.uri.Uri
import reactor.core.publisher.Flux
import java.io.Serializable

/**
 * Resource type e.g. Table, Index, Stored function etc
 */
interface ResourceType : Serializable {

    /**
     * Name of resource type, e.g "table", "index", "stored procedure" etc
     */
    fun name() = aliases()[0]

    /**
     * Aliases of resource type, e.g. ["table", "tbl", "t", "tablo"]
     */
    fun aliases(): List<String>

    /**
     * Path of resource e.g. ["schema", "table", "column"] or ["schema", "table", "index", "index_column"]
     */
    fun path(): List<String>

    /**
     * Information about the resource metadata (properties) "name" -> "type"
     * TODO rename to columns(uri) ?
     */
    fun metaData(uri: Uri): List<Metadata>

    /**
     * Read resources by uri
     */
    fun read(uri: Uri): Flux<DbResource>

    /**
     * Get as JSON
     */
    fun toJson(): JsonNode

}
